src/test/java/com/atlassian/db/replica/it/DualConnectionIT.java

```diff
@@ -20,4 +25,63 @@ public void shouldUseReplica() throws SQLException {
             }
         }
     }
+
+
+    @Test
+    public void shouldPreserveAutoCommitModeWhileSwitchingFromReplicaToMain() throws SQLException {
+        try (PostgresConnectionProvider connectionProvider = new PostgresConnectionProvider()) {
+            final Connection connection = DualConnection.builder(
+                connectionProvider,
+                new CircularConsistency.Builder(ImmutableList.of(false, true)).build()
+            ).build();
+
+            connection.setAutoCommit(false);
+            connection.prepareStatement("SELECT 1;").executeQuery();
+            connection.prepareStatement("SELECT 1;").executeQuery();
+            connection.commit();
+        }
+    }
+
+    @Test
+    public void shouldRunNextValOnMainDatabase() throws SQLException {
+        try (PostgresConnectionProvider connectionProvider = new PostgresConnectionProvider()) {
+            createTestSequence(connectionProvider);
+
+            final Connection connection = DualConnection.builder(
+                connectionProvider,
+                CircularConsistency.permanentConsistency().build()
+            ).build();
+
+            connection.prepareStatement("SELECT nextval('test_sequence');").executeQuery();
+        }
+    }
+
+    private void createTestSequence(PostgresConnectionProvider connectionProvider) throws SQLException {
+        final Connection mainConnection = connectionProvider.getMainConnection();
+        try (final Statement mainStatement = mainConnection.createStatement()) {
+            mainStatement.execute("CREATE SEQUENCE test_sequence;");
+            final LsnReplicaConsistency consistency = new LsnReplicaConsistency();
+            waitForReplicaConsistency(connectionProvider, mainConnection, consistency);
+        }
+    }
+
+    private void waitForReplicaConsistency(
+        PostgresConnectionProvider connectionProvider,
+        Connection mainConnection,
+        LsnReplicaConsistency consistency
+    ) {
+        consistency.write(mainConnection);
+        for (int i = 0; i < 30; i++) {
+            if (consistency.isConsistent(connectionProvider::getReplicaConnection)) {
+                return;
+            }
+            try {
+                Thread.sleep(1000);
```

Comment:
Would it generate some flakiness in the future? It's a good trade-off for now, though it's nto sth like `CountDownLatch`

Reply:
You mean spurious wake-ups?

Reply:
This is a simple impl of a timeout. As an integration test, a real `Clock.systemUTC()` is used, so we need time to pass. It's not like we can inject a latch into various participants and make them cooperate

Reply:
I wrote an integration test to be sure I'm solving the same issue we see on prod. I plan to rewrite it as a unit test.

Reply:
Fair enough, below convinced me, thanks for explaining the reasoning behind trade-off. 👍 
> As an integration test, a real Clock.systemUTC() is used